File: backend/app/schemas/m2.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.m1 import ChunkingConfig

DocumentStatus = Literal["pending", "running", "completed", "failed"]
TaskStatus = Literal["pending", "running", "completed", "failed"]
TaskType = Literal["document_process", "wiki_ingest", "wiki_rebuild"]
TaskStage = Literal[
    "pending",
    "chunking",
    "embedding",
    "indexing",
    "extracting",
    "citing",
    "taxonomy",
    "summarizing",
    "reducing",
    "postprocessing",
    "completed",
    "failed",
]
ChunkType = Literal["text", "wiki_page"]


class TagOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    kb_id: str
    name: str


class TagCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=64)


class ChunkOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    document_id: str | None = None
    kb_id: str
    content: str
    header_path: list[str]
    seq: int
    start_pos: int
    end_pos: int
    chunk_type: ChunkType
    source_page_id: str | None = None
    created_at: datetime


class DocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    kb_id: str
    filename: str
    file_hash: str
    file_size: int
    status: DocumentStatus
    error_message: str | None = None
    chunk_count: int
    tags: list[TagOut] = []
    created_by: str
    created_by_username: str
    created_at: datetime
    updated_at: datetime


class DocumentPage(BaseModel):
    items: list[DocumentOut]
    total: int
    page: int
    page_size: int


class DocumentDetail(DocumentOut):
    content: str
    chunks: list[ChunkOut]


class DocumentUploadResponse(BaseModel):
    documents: list[DocumentOut]
    task_ids: list[str]


class ChunkPreviewRequest(BaseModel):
    content: str = Field(min_length=1)
    content_type: Literal["markdown", "text"]
    chunking_config: ChunkingConfig | None = None

    @model_validator(mode="after")
    def validate_overlap(self) -> "ChunkPreviewRequest":
        config = self.chunking_config
        if config is not None and config.chunk_overlap >= config.chunk_size:
            raise ValueError("chunk_overlap must be smaller than chunk_size")
        return self


class ChunkPreviewItem(BaseModel):
    content: str
    header_path: list[str]
    seq: int
    start_pos: int
    end_pos: int
    char_count: int


class ChunkPreviewResponse(BaseModel):
    items: list[ChunkPreviewItem]


class TaskAcceptedResponse(BaseModel):
    task_id: str


class TaskError(BaseModel):
    code: str
    message: str
    details: dict[str, Any] = {}


class TaskOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    kb_id: str
    task_type: TaskType
    status: TaskStatus
    stage: TaskStage
    progress: int
    payload: dict[str, Any]
    run_after: datetime | None = None
    dedup_key: str | None = None
    error: TaskError | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/src/api/m2.ts
import { apiClient } from "./client";

export type DocumentStatus = "pending" | "running" | "completed" | "failed";

export type Tag = {
  id: string;
  kb_id: string;
  name: string;
};

export type Document = {
  id: string;
  kb_id: string;
  filename: string;
  file_hash: string;
  file_size: number;
  status: DocumentStatus;
  error_message?: string | null;
  chunk_count: number;
  tags?: Tag[];
  created_by: string;
  created_by_username: string;
  created_at: string;
  updated_at: string;
};

export type Chunk = {
  id: string;
  document_id?: string | null;
  kb_id: string;
  content: string;
  header_path: string[];
  seq: number;
  start_pos: number;
  end_pos: number;
  chunk_type: "text" | "wiki_page";
  source_page_id?: string | null;
  created_at: string;
};

export type DocumentDetail = Document & {
  content: string;
  chunks: Chunk[];
};

export type Task = {
  id: string;
  kb_id: string;
  task_type: "document_process" | "wiki_ingest" | "wiki_rebuild";
  status: "pending" | "running" | "completed" | "failed";
  stage: string;
  progress: number;
  payload: Record<string, unknown>;
  error?: { code: string; message: string; details?: Record<string, unknown> } | null;
  created_at: string;
  updated_at: string;
};

export type ChunkPreviewItem = {
  content: string;
  header_path: string[];
  seq: number;
  start_pos: number;
  end_pos: number;
  char_count: number;
};

export async function listDocuments(params: {
  kbId: string;
  q?: string;
  tag_id?: string;
  status?: DocumentStatus | "";
  sort?: string;
}): Promise<Document[]> {
  const { kbId, ...query } = params;
  const response = await apiClient.get<{ items: Document[] }>(`/kbs/${kbId}/documents`, { params: query });
  return response.data.items;
}

export async function getDocument(documentId: string): Promise<DocumentDetail> {
  const response = await apiClient.get<DocumentDetail>(`/documents/${documentId}`);
  return response.data;
}

export async function uploadDocuments(payload: {
  kbId: string;
  files: File[];
  tagIds: string[];
}): Promise<{ documents: Document[]; task_ids: string[] }> {
  const form = new FormData();
  payload.files.forEach((file) => form.append("files", file));
  payload.tagIds.forEach((tagId) => form.append("tag_ids", tagId));
  const response = await apiClient.post<{ documents: Document[]; task_ids: string[] }>(`/kbs/${payload.kbId}/documents/upload`, form);
  return response.data;
}

export async function deleteDocument(documentId: string): Promise<void> {
  await apiClient.delete(`/documents/${documentId}`);
}

export async function retryDocument(documentId: string): Promise<{ task_id: string }> {
  const response = await apiClient.post<{ task_id: string }>(`/documents/${documentId}/retry`);
  return response.data;
}

export async function listTags(kbId: string): Promise<Tag[]> {
  const response = await apiClient.get<{ items: Tag[] }>(`/kbs/${kbId}/tags`);
  return response.data.items;
}

export async function createTag(kbId: string, name: string): Promise<Tag> {
  const response = await apiClient.post<Tag>(`/kbs/${kbId}/tags`, { name });
  return response.data;
}

export async function updateTag(kbId: string, tagId: string, name: string): Promise<Tag> {
  const response = await apiClient.patch<Tag>(`/kbs/${kbId}/tags/${tagId}`, { name });
  return response.data;
}

export async function deleteTag(kbId: string, tagId: string): Promise<void> {
  await apiClient.delete(`/kbs/${kbId}/tags/${tagId}`);
}

export async function getTask(taskId: string): Promise<Task> {
  const response = await apiClient.get<Task>(`/tasks/${taskId}`);
  return response.data;
}

export async function previewChunks(payload: {
  kbId: string;
  content: string;
  content_type: "markdown" | "text";
  chunking_config?: { chunk_size: number; chunk_overlap: number; strategy: "header_aware" };
}): Promise<ChunkPreviewItem[]> {
  const { kbId, ...body } = payload;
  const response = await apiClient.post<{ items: ChunkPreviewItem[] }>(`/kbs/${kbId}/chunk-preview`, body);
  return response.data.items;
}
